lib/src/client.ts

```diff
@@ -27,220 +26,243 @@ export interface HeadersObject {
   }
 
-  return parseCommit(body);
-}
+  /**
+   * Fetches and Parses a Resource. Can fetch through another atomic server if you
+   * pass the `from` argument, which should be the baseURL of an Atomic Server.
+   * Returns a tuple of the requested resource and a list of all resources found in the response.
+   */
+  public async fetchResourceHTTP(
+    subject: string,
+    opts: FetchResourceOptions = {},
+  ): Promise<HTTPResult> {
+    const { signInfo, from } = opts;
+    let createdResources: Resource[] = [];
+    const parser = new JSONADParser();
+    let resource = new Resource(subject);
+
+    try {
+      Client.tryValidURL(subject);
+      const requestHeaders: HeadersObject = {};
+      requestHeaders['Accept'] = JSON_AD_MIME;
+
+      if (signInfo) {
+        // Cookies only work for same-origin requests right now
+        // https://github.com/atomicdata-dev/atomic-data-browser/issues/253
+        if (subject.startsWith(window.location.origin)) {
+          if (!checkAuthenticationCookie()) {
+            setCookieAuthentication(signInfo.serverURL, signInfo.agent);
+          }
+        } else {
+          await signRequest(subject, signInfo.agent, requestHeaders);
+        }
+      }
 
-/** Throws an error if the URL is not valid */
-export function tryValidURL(subject: string | undefined): void {
-  try {
-    new URL(subject as string);
-  } catch (e) {
-    throw new Error(`Not a valid URL: ${subject}. ${e}`);
-  }
-}
+      let url = subject;
 
-/** Returns true if a URL is valid. */
-export function isValidURL(subject: string | undefined): boolean {
-  if (typeof subject !== 'string') return false;
+      if (from !== undefined) {
+        const newURL = new URL(`${from}/path`);
+        newURL.searchParams.set('path', subject);
+        url = newURL.href;
+      }
+
+      const response = await this.fetch(url, {
+        headers: requestHeaders,
+      });
+      const body = await response.text();
+
+      if (response.status === 200) {
+        try {
+          const json = JSON.parse(body);
+
+          const [parsedResource, parsedCreatedResources] = parser.parseObject(
+            json,
+            subject,
+          );
+
+          resource = parsedResource;
+          createdResources.push(...parsedCreatedResources);
+        } catch (e) {
+          throw new AtomicError(
+            `Could not parse JSON from fetching ${subject}. Is it an Atomic Data resource? Error message: ${e.message}`,
+          );
+        }
+      } else if (response.status === 401) {
+        throw new AtomicError(body, ErrorType.Unauthorized);
+      } else if (response.status === 500) {
+        throw new AtomicError(body, ErrorType.Server);
+      } else if (response.status === 404) {
+        throw new AtomicError(body, ErrorType.NotFound);
+      } else {
+        throw new AtomicError(body);
+      }
+    } catch (e) {
+      resource.setError(e);
+      createdResources = [resource];
+      console.error(subject, e);
```

Comment:
I think this should use `store.handleError`, but I suppose that is not available here, right?

Reply:
I didn't know that was a thing. Shouldn't be the concern of Client to handle tho. `store.fetchResourceFromServer` can check if the returned resource has an error and pass that to the errorHandler.

After looking at errorHandler tho I think the name is misleading as it doesn't actually handle the error most of the time but just notifies the user the store encountered a problem. We could replace it with an OnError event on the already existing eventManager, this way it is also inline with other pub sub things in the store.

Reply:
Agreed